### PR TITLE
Baremetal support

### PR DESCRIPTION
**Pull request by sleffler**
This is the baremetal support I'm using to build OpenTitan and OT-based applications (available at opensecura.googlesource.com). These are tested in simulation.

There are XXX comments that are reminders for me; will remove after resolving feedback.

Include files are derived from cheriot-rtos; I've preserved the copyright.

### CODE REVIEW

File: clang/lib/Basic/Targets/RISCV.h [hunk 1]
```diff
@@ -87,6 +91,7 @@ class RISCVTargetInfo : public TargetInfo {
   bool HasZvamo = false;
   bool HasZvlsseg = false;
   bool IsABICHERIoT = false;
+  bool IsCHERIoTBareMetal = false;
```
Comment (vmurali) [inline clang/lib/Basic/Targets/RISCV.h:94]: Bike shedding: IsABICHERIoTBareMetal instead?

File: clang/lib/Basic/Targets/RISCV.h [hunk 2]
```diff
@@ -181,11 +187,17 @@ class RISCVTargetInfo : public TargetInfo {
...
+    if (CC == CallingConv::CC_CHERICCall) {
+      // NB: suppress warning when cheriot-baremetal
+      return ABI == "cheriot" || ABI == "cheriot-baremetal" ? CCCR_OK : CCCR_Warning;
+    }
+    if ((CC == CallingConv::CC_CHERICCallee) ||
         (CC == CallingConv::CC_CHERICCallback) ||
-        (CC == CallingConv::CC_CHERILibCall))
+        (CC == CallingConv::CC_CHERILibCall)) {
+      // NB: with cheriot-baremetal the caller will generate a warning
+      //   and downgrade to the default target CC
```
Comment (vmurali) [inline clang/lib/Basic/Targets/RISCV.h]: Not sure I understand the need for this change; you never generate a CC_CHERICCall, right, for baremetal?

File: llvm/include/llvm/Support/Base64.h [hunk 3]
```diff
@@ -13,6 +13,7 @@
 #ifndef LLVM_SUPPORT_BASE64_H
 #define LLVM_SUPPORT_BASE64_H
 
+#include <cstdint>
```
Comment (vmurali) [inline llvm/include/llvm/Support/Base64.h]: Why is this needed?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -8632,9 +8639,11 @@ SDValue RISCVTargetLowering::LowerCall(CallLoweringInfo &CLI,
     if ((CallConv == CallingConv::CHERI_CCall) ||
         (CallConv == CallingConv::CHERI_CCallee))
       Chain = DAG.getNode(RISCVISD::CAP_COMPARTMENT_CALL, DL, NodeTys, Ops);
-    else if (CallConv == CallingConv::CHERI_LibCall)
+    else if (CallConv == CallingConv::CHERI_LibCall) {
+      assert(!Subtarget.getTargetABI() != RISCVABI::ABI_CHERIOT_BAREMETAL &&
```
Comment (vmurali) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Can we remove the double negative?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 5]
```diff
@@ -75,10 +76,13 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
     ABI = Subtarget.is64Bit() ? RISCVABI::ABI_LP64 : RISCVABI::ABI_ILP32;
   }
 
-  if (ABI == RISCVABI::ABI_CHERIOT)
+  if (ABI == RISCVABI::ABI_CHERIOT) {
     for (int LCID = 0; LCID < RTLIB::UNKNOWN_LIBCALL; ++LCID)
-      setLibcallCallingConv(static_cast<RTLIB::Libcall>(LCID),
-                            CallingConv::CHERI_LibCall);
+      setLibcallCallingConv(static_cast<RTLIB::Libcall>(LCID), CallingConv::CHERI_LibCall);
```
Comment (vmurali) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Whitespace change

File: clang/lib/Baremetal/CMakeLists.txt
Comment (vmurali) [inline clang/lib/Baremetal/CMakeLists.txt:1]: Do we need a separate directory for Baremetal? The only change seems to be redefining __cheri_*. And that is already protected by the _CHERIOT_BAREMETAL_ macro. The compilation path could be the same as CHERIoT's default, except for using CCalls instead of CHERI_CCalls.

Comment (davidchisnall) [reply clang/lib/Baremetal/CMakeLists.txt:1]: No, the clang headers are expected to work in all configurations.  If we want clang to provide headers, we should fix the ones that clang provides to work with CHERIoT.  I don't really want to do this, since we don't install the clang headers by default in the CHERIoT toolchain builds, I'd much rather that we change the RTOS headers to work with bare metal (which, I believe, requires a trivial change in `cdefs.h` to define the `__cheri_libcalls` and `__cheri_compartment` macros to nothing when in bare-metal mode.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 6]
```diff
@@ -75,10 +76,13 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
...
+  } else if (ABI == RISCVABI::ABI_CHERIOT_BAREMETAL) {
+    for (int LCID = 0; LCID < RTLIB::UNKNOWN_LIBCALL; ++LCID)
+      setLibcallCallingConv(static_cast<RTLIB::Libcall>(LCID), CallingConv::CHERI_CCall);
+  }
```
Comment (davidchisnall) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Is this right?  I don't think we want libcalls to use the cross-compartment calling convention in bare-metal mode.

Comment (vmurali) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: +1 CallingConv::CC_C seems to be the correct one?

Comment (davidchisnall) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Yup, and that’s the default if you don’t set anything explicitly.

File: clang/lib/Baremetal/CMakeLists.txt
Comment (sleffler) [reply clang/lib/Baremetal/CMakeLists.txt:1]: I removed the baremetal subdir. I think it's a bad idea to tie baremetal compilation to cheriot-rtos (e.g. there are other potential os's that might be used with the toolchain). But I'm also not thrilled that install-baremetal-headers and install-clang-resource-headers overlap though it should be a noop (unless something depends on the file mod times).

I can remove the reach-over includes and just require the toolchain user install clang-resource-headers to get them.

**Review by davidchisnall — APPROVED** [review]

File: clang/lib/Baremetal/CMakeLists.txt
Comment (sleffler) [reply clang/lib/Baremetal/CMakeLists.txt:1]: Removed headers for now based on private discussion.
